File: heaptest/heap.c

#include <stdio.h>
#include <stdlib.h>

struct Elem {
    int key;
    void* value;
};

#define HEAP_SIZE 100

struct Elem** _buf;
int _size;

void swap(int i, int j)
{
    struct Elem* temp = _buf[i];
    _buf[i] = _buf[j];
    _buf[j] = temp;
}

void heapify(int i)
{
    int p = (i-1) >> 1; // (int)floor((i-1)/2);

    int l = (i << 1) + 1; // 2 * i + 1
    int r = (i << 1) + 2; // 2 * i + 2

    struct Elem* e = _buf[i];

    if (i != 0) { // parent check
        struct Elem* parent = _buf[p];
        if (e->key < parent->key) {
            swap(i, p);
            heapify(p);
        }
    }

    if (_size >= l) { // child check
        if (_size <= r) { // only have left
            struct Elem* left = _buf[l];
            if (e->key > left->key) {
                swap(i, l);
                heapify(l);
            }
        }
        else { // have left and right
            struct Elem* left = _buf[l];
            struct Elem* right = _buf[r];
            int s = left->key < right->key ? l : r;
            struct Elem* child = _buf[s];
            if (e->key > child->key) {
                swap(i, s);
                heapify(s);
            }
        }
    }
}

void init()
{
    _buf = calloc(HEAP_SIZE, sizeof(struct Elem));
    _size = 0;
}

void uninit()
{
    for (int i = 0; i < _size; ++i) {
        free(_buf[i]);
    }
    free(_buf);
}

struct Elem* findMin()
{
    return _buf[0];
}

struct Elem* extractMin()
{
    struct Elem* min = findMin();
    _buf[0] = _buf[_size-1];
    _size -= 1;
    heapify(0);

    return min;
}

void insert(int key, void* value)
{
    struct Elem* elem = malloc(sizeof(struct Elem));
    elem->key = key;
    elem->value = value;
    _buf[_size] = elem;
    heapify(_size);
    _size += 1;
}

void changeKey(int index, int newKey)
{
    _buf[index]->key = newKey;
    heapify(index);
}

void printheap()
{
    for (int i = 0; i < _size; ++i) {
        int* pVal = _buf[i]->value;
        printf("(%d => %d), ", _buf[i]->key, *pVal);
    }
    printf("\n");
}

int main(int argc, char* argv[])
{
    int val = 1;

    init();

    insert(2, &val);
    printheap();

    insert(2, &val);
    printheap();

    insert(13, &val);
    printheap();

    insert(7, &val);
    printheap();

    insert(3, &val);
    printheap();

    insert(-1, &val);
    printheap();

    printf("Change Key\n");
    changeKey(0, 14);
    printheap();

    insert(77, &val);
    printheap();

    insert(8, &val);
    printheap();

    insert(0, &val);
    printheap();

    int items = _size;
    for (int i = 0; i < items; ++i) {
        struct Elem* e = extractMin();
        printf("Extract Min: %d\n", e->key);
        printheap();
    }

    uninit();

    return 0;
}

File: heaptest/heap.py
#!/usr/bin/env python

class Heap:
    "Heap Abstract Data Type"

    def __init__(self, L=None):
        self.data = []
        if L:
            for e in L:
                self.insert(e[0], e[1])

    def __len__(self):
        return len(self.data)

    def __str__(self):
        return ', '.join(["[%s: %s]" % (d[0], d[1]) for d in self.data])

    def findmin(self):
        """Return min element from this heap without removing it. Element
        is an array of size 2 with key at first index and value at second:
        [key, val]"""
        try:
            return self.data[0]
        except:
            raise HeapException, "findmin: Heap empty"

    def extractmin(self):
        """Find and remove min element from this heap. Element is an array
        of size 2 with key at first index and value at second: [key, val]"""
        try:
            if len(self.data) == 1:
                return self.data.pop()
            else:
                m = self.findmin()
                self.data[0] = self.data.pop()
                self._heapify(0)

                return m
        except:
            raise HeapException, "extractmin: Heap empty"

    def insert(self, key, val):
        """Insert key/value pair into this heap. Key/value pair should be
        passed as 2-element array: [key, val]"""
        self.data.append([key, val])
        # Subtract 1 for heapify since we are zero-based
        self._heapify(len(self.data) - 1)

    def changekey(self, index, newkey):
        # Cur key at index 0 of list at self.data[index]
        self.data[index][0] = newkey
        self._heapify(index)

    def _heapify(self, i):
        p = (i - 1) >> 1    # floor((i - 1) / 2)
        l = (i << 1) + 1    # i * 2 + 1
        r = (i << 1) + 2    # i * 2 + 2

        e = self.data[i]

        if i != 0:
            parent = self.data[p]
            if e[0] < parent[0]:
                self._swap(i, p)
                self._heapify(p)

        if len(self.data) > l:          # Check for children
            if len(self.data) - 1 < r:  # Only have left
                left = self.data[l]
                if e[0] > left[0]:
                    self._swap(i, l)
                    self._heapify(l)
            else:                       # Have left and right
                left = self.data[l]
                right = self.data[r]
                if left[0] < right[0]:
                    s = l
                else:
                    s = r
                child = self.data[s]
                if e[0] > child[0]:
                    self._swap(i, s)
                    self._heapify(s)

    def _swap(self, i, j):
        temp = self.data[i]
        self.data[i] = self.data[j]
        self.data[j] = temp


class HeapException(Exception):
    pass


if __name__ == '__main__':
    P = Heap([[2, 1], [2, 1], [13, 1], [7, 1], [3, 1], [-1, 1]])
    print P

    print "Change Key"
    P.changekey(0, 14)
    print P

    P.insert(77, 1)
    print P

    P.insert(8, 1)
    print P

    P.insert(0, 1)
    print P

    while P:
        print "Extract Min: %s" % (P.extractmin()[0])
        print P
